langchain4j-vertex-ai/src/main/java/dev/langchain4j/model/vertexai/VertexAiMatchingEngine.java

```diff
@@ -0,0 +1,290 @@
+
+    /**
+     * Adds all the embeddings to the index.
+     *
+     * @param embeddings A list of embeddings to be added to the store.
+     * @return the ids of the embeddings
+     */
+    @Override
+    public List<String> addAll(List<Embedding> embeddings) {
+        return addAll(null, embeddings, null);
+    }
+
+    /**
+     * Adds all the embeddings to the index.
+     *
+     * @param embeddings A list of embeddings to be added to the store.
+     * @param embedded   A list of text segments to be added to the store.
+     * @return the ids of the embeddings
+     */
+    @Override
+    public List<String> addAll(List<Embedding> embeddings, List<TextSegment> embedded) {
+        return addAll(null, embeddings, embedded);
+    }
+
+    /**
+     * Adds all the embeddings to the index.
+     *
+     * @param ids        the ids of the embeddings
+     * @param embeddings A list of embeddings to be added to the store.
+     * @param embedded   A list of text segments to be added to the store.
+     * @return the ids of the embeddings
+     */
+    public List<String> addAll(List<String> ids, List<Embedding> embeddings, List<TextSegment> embedded) {
+        if (isCollectionEmpty(embeddings)) {
+            log.info("Empty embeddings - no ops");
+            return new ArrayList<>();
+        }
+        ensureTrue(embedded == null || embedded.size() == embeddings.size(), "embeddings size is not equal to embedded size");
+        ensureTrue(ids == null || ids.size() == embeddings.size(), "ids size is not equal to embeddings size");
+
+        final VertexAiEmbeddingIndex index = new VertexAiEmbeddingIndex();
+        final List<String> insertedIds = new ArrayList<>();
+        final String indexFileId = randomUUID();
+
+        for (int embeddingIndex = 0; embeddingIndex < embeddings.size(); embeddingIndex++) {
+            final Embedding embedding = embeddings.get(embeddingIndex);
+
+            TextSegment textSegment;
+            String id;
+            if (embedded != null) {
+                textSegment = embedded.get(embeddingIndex);
+                id = (ids != null)
+                        ? ids.get(embeddingIndex)
+                        : (avoidDups ? generateUUIDFrom(textSegment.text()) : randomUUID());
+            } else {
+                textSegment = null;
+                id = (ids != null)
+                        ? ids.get(embeddingIndex)
+                        : randomUUID();
+            }
+
+            // Add the embedding to the index
+            index.addEmbedding(id, embedding);
+            insertedIds.add(id);
+
+            // Upload the document to GCS
+            getGcpBlobService().upload("documents/" + id, new VertexAIDocument(indexFileId, textSegment));
+        }
+
+        final String filename = "indexes/" + indexFileId + ".json";
+        log.info("Uploading {} index to GCS.", filename);
+
+        // Upload the index to GCS
+        getGcpBlobService().upload(filename, index.toString());
```

Comment:
is it really required to upload the index (embeddings) both into a bucket AND index endpoint?

Reply:
For the documentation yes and looked at the LangChain python version of and they also do the same.

Reply:
[LangChain implementation](https://github.com/langchain-ai/langchain/blob/ee94ef55ee6ab064da08340817955f821dfa6261/libs/langchain/langchain/vectorstores/matching_engine.py#L159) seem to upload embeddings to GCS bucket first and then update the index, providing a reference to the bucket (where to read embeddings from). But in your implementation you provide embeddings directly in the index, so it feels that uploading them to GCS is unnecessary. Could you please remove this line and test if it still works? Thanks!